src/node_contextify.cc

```diff
@@ -1497,73 +1482,89 @@ void ContextifyContext::ContainsModuleSyntax(
 
   bool should_retry_as_esm = false;
   if (try_catch.HasCaught() && !try_catch.HasTerminated()) {
-    Utf8Value message_value(env->isolate(), try_catch.Message()->Get());
-    auto message = message_value.ToStringView();
+    should_retry_as_esm =
+        ContextifyContext::ShouldRetryAsESMInternal(env, code);
+  }
+  args.GetReturnValue().Set(should_retry_as_esm);
+}
 
-    for (const auto& error_message : esm_syntax_error_messages) {
-      if (message.find(error_message) != std::string_view::npos) {
-        should_retry_as_esm = true;
-        break;
-      }
-    }
+void ContextifyContext::ShouldRetryAsESM(
+    const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+
+  CHECK_EQ(args.Length(), 1);
+
+  // Argument 1: source code
+  Local<String> code;
+  CHECK(args[0]->IsString());
+  code = args[0].As<String>();
+
+  bool should_retry_as_esm =
+      ContextifyContext::ShouldRetryAsESMInternal(env, code);
+
+  args.GetReturnValue().Set(should_retry_as_esm);
```

Comment:
Nit

We generally add a comment to CHECK_EQ(length) the parameters as the comment. For example `node_file.cc`

```suggestion
  CHECK_EQ(args.Length(), 1); // code
  CHECK(args[0]->IsString());

  bool should_retry_as_esm =
      ContextifyContext::ShouldRetryAsESMInternal(env, args[0].As<String>());
  args.GetReturnValue().Set(should_retry_as_esm);
```